test/object_store/test_basic.py

```diff
@@ -53,9 +53,14 @@ def run_with_dir(run_cmd_gen, run_dir):
         log = os.path.join(run_dir, 'log')
         log_fd = os.open(log, os.O_WRONLY | os.O_CREAT | os.O_APPEND, mode=0o666)
         # redirect stdout and stderr to log file
-        for output in [sys.stdout, sys.stderr]:
+        # we have to hardwire the STDOUT_FILENO and STDERR_FILENO. because
+        # pytest mutates sys.stdout and sys.stderr if it captures them, so,
+        # if we intent to redirect the child process's stdout and stderr fd
+        # to the specified fd, we have to use the fd numbers *used* by the
+        # child process, not the ones used by the test.
+        outputs = [(sys.stdout, 1), (sys.stderr, 2)]
+        for output, output_fd in outputs:
             output.flush()
-            output_fd = output.fileno()
```

Comment:
Interesting. sys.stdout.fileno() is supposed to be 1, who changed it to be anything else?

I just noticed that Python also has sys.__stdout__ et al, and those are supposedly the original stdout (fd 1) and friends. Supposedly, a Python program can assign sys.stdout... But who did this? Does test.py to this? :-(

So I think you can indeed hard-code the numbers 1,2 or use sys.__stdout and then fileno() will work as hoped. I think.

Reply:
> Interesting. sys.stdout.fileno() is supposed to be 1, who changed it to be anything else?

please refer to the commit message. i am quoting part of it here

>> pytest changes the test's stdout and stderr to the captured
fds when it captures the outputs of the test.

if you are curious no pytest, please see https://github.com/pytest-dev/pytest/blob/b73b4c464c6fa54f55a1c6ed8eaceb5529a161bb/src/_pytest/capture.py#L129-L131 . FWIW, it even has some fixtures allowing the tests to access the real stdout and stderr.
> 
> I just noticed that Python also has sys.**stdout** et al, and those are supposedly the original stdout (fd 1) and friends. Supposedly, a Python program can assign sys.stdout... But who did this? Does test.py to this? :-(
> 

explained above.

> So I think you can indeed hard-code the numbers 1,2 or use sys.__stdout and then fileno() will work as hoped. I think.

thank you! TIL. good point. will use `sys.__stdout__`, i am not a big fan of magic numbers in any form.

Reply:
But `sys.__stdout__` is documented to be some obscure "original stdout". It's the stdout as current process sees them or as parent processes decided them to be. Opposite to that, our goal is to close and redirect the stdouts _as the execved process_ wants them regardless of what current/parent/python things of them, so it _must_ be hand-coded 1 and 2

Reply:
fair enough. but we design by contract, if someone breaks the protocol, that's a bug. the protocol here is the "original stdout". if it is "obscure", it is a bug as well.

i will change to use 1 and 2 back to in the first revision. as it does not worth the efforts to do the bike shedding for removing the magic numbers, i think.